File: src/main/kotlin/broker/ConnectionsManager.kt
package broker

import akka.actor.AbstractActor
import akka.actor.ActorRef
import akka.actor.Props
import akka.actor.Terminated
import akka.http.javadsl.model.ws.Message

class ConnectionsManager : AbstractActor() {

    object Join

    var connections = emptySet<ActorRef>()

    override fun createReceive(): Receive {
        return receiveBuilder()
            .match(Join::class.java) {
                connections = connections + sender
                context.watch(sender)
            }
            .match(Terminated::class.java) {
                connections = connections - it.actor
            }
            .match(Message::class.java) { msg ->
                connections.forEach { conn ->
                    if (conn != sender) {
                        conn.tell(msg, self)
                    }
                }
            }
            .build()
    }

    companion object {
        fun props(): Props =
            Props.create(ConnectionsManager::class.java) { ConnectionsManager() }
    }
}

File: src/main/scala/com/quicks/broker/ConnectionsManager.scala
package com.quicks.broker

import akka.actor.{Actor, ActorLogging, ActorRef, Terminated}
import akka.http.scaladsl.model.ws.Message

object Join

class ConnectionsManager extends Actor with ActorLogging {

  var connections: Set[ActorRef] = Set.empty

  override def receive = {

    case Join =>
      log.info("New connection joined !")
      connections += sender()
      context.watch(sender())

    case Terminated(conn) =>
      log.info("Connection was terminated !")
      connections -= conn

    case m: Message =>
      log.info("Broadcasting message !")
      connections.foreach( conn => {
        if (conn != sender()) {
          conn ! m
        }
      })

  }

}
